drivers/storage/port/scsiport/fdo.c

```diff
@@ -589,106 +503,249 @@ CallHWInitialize(
-        ExFreePool(DeviceExtension->SrbInfo);
+        ExFreePoolWithTag(DeviceExtension->SrbInfo, TAG_SCSIPORT);
 
     /* Unmap mapped addresses */
     while (DeviceExtension->MappedAddressList != NULL)
     {
         MmUnmapIoSpace(DeviceExtension->MappedAddressList->MappedAddress,
                        DeviceExtension->MappedAddressList->NumberOfBytes);
 
-        Ptr = DeviceExtension->MappedAddressList;
+        PVOID ptr = DeviceExtension->MappedAddressList;
         DeviceExtension->MappedAddressList = DeviceExtension->MappedAddressList->NextMappedAddress;
 
-        ExFreePool(Ptr);
+        ExFreePoolWithTag(ptr, TAG_SCSIPORT);
     }
 
-    /* Finally delete the device object */
-    DPRINT("Deleting device %p\n", DeviceExtension->DeviceObject);
     IoDeleteDevice(DeviceExtension->DeviceObject);
+
+    return STATUS_SUCCESS;
+}
+
+NTSTATUS
+FdoStartAdapter(
+    _In_ PSCSI_PORT_DEVICE_EXTENSION PortExtension)
+{
+    WCHAR dosNameBuffer[12];
+    UNICODE_STRING dosDeviceName;
+    NTSTATUS status;
+
+    // Start our timer
+    IoStartTimer(PortExtension->DeviceObject);
+
+    // Create the dos device link
+    swprintf(dosNameBuffer, L"\\??\\Scsi%u:", PortExtension->PortNumber);
+    RtlInitUnicodeString(&dosDeviceName, dosNameBuffer);
+    status = IoCreateSymbolicLink(&dosDeviceName, &PortExtension->DeviceName);
+    if (!NT_SUCCESS(status))
+    {
+        return status;
+    }
+
+    // start building a device map
+    RegistryInitAdapterKey(PortExtension);
+
+    // increase the port count
+    PCONFIGURATION_INFORMATION sysConfig = IoGetConfigurationInformation();
+    sysConfig->ScsiPortCount++;
+
+    // Register and enable the device interface
+    status = IoRegisterDeviceInterface(PortExtension->DeviceObject,
+                                       &StoragePortClassGuid,
+                                       NULL,
+                                       &PortExtension->InterfaceName);
+    DPRINT("IoRegisterDeviceInterface status: %x, InterfaceName: %wZ\n",
+        status, &PortExtension->InterfaceName);
+
+    if (NT_SUCCESS(status))
+    {
+        IoSetDeviceInterfaceState(&PortExtension->InterfaceName, TRUE);
+    }
+
+    PortExtension->DeviceStarted = TRUE;
+
+    return STATUS_SUCCESS;
+}
+
+static
+NTSTATUS
+FdoHandleDeviceRelations(
+    _In_ PSCSI_PORT_DEVICE_EXTENSION PortExtension,
+    _Inout_ PIRP Irp)
+{
+    PIO_STACK_LOCATION ioStack = IoGetCurrentIrpStackLocation(Irp);
+
+    // FDO always only handles bus relations
+    if (ioStack->Parameters.QueryDeviceRelations.Type == BusRelations)
+    {
+        FdoScanAdapter(PortExtension);
+        DPRINT("Found %u PD objects, FDOExt: %p\n", PortExtension->TotalLUCount, PortExtension);
+
+        PDEVICE_RELATIONS deviceRelations =
+            ExAllocatePoolWithTag(PagedPool,
+                                  (sizeof(DEVICE_RELATIONS) +
+                                   sizeof(PDEVICE_OBJECT) * (PortExtension->TotalLUCount - 1)),
```

Comment:
`DEVICE_RELATIONS` structure has 1 `PDEVICE_OBJECT` element inside it. So in case `TotalLUCount` is 0, it will allocate memory as if there are 0 elements in the array.
ReactOS will handle such case, but I haven't checked on Windows